src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -1,69 +1,141 @@
-        return;
-    }
+  public User getUser(String userId) {
+    return null;
+  }
+  public String addUser() {
+    return null;
+  }
+  public void deleteUser(String userId) {
+    return;
+  }
+  public void addRecipeIdToCreated(String userId, String recipeId) {
+    return;
+  }
+  public void addRecipeIdToSaved(String userId, String recipeId) {
+    return;
+  }
+  public void followTag(String userId, String tagId) {
+    return;
+  }
 
-    public Iterable<RecipeMetadata> getRecipesMatchingTags(Iterable<String> tagIds) {
-        return null;
-    }
-    public Iterable<RecipeMetadata> getRecipesMatchingCreator(String creatorId) {
-        return null;
-    }
-    public Iterable<RecipeMetadata> getRecipesMatchingIDs(Iterable<String> Ids) {
-        return null;
-    }
+  public Iterable<RecipeMetadata> getRecipesMatchingTags(Iterable<String> tagIds) {
+    Query query = recipesMatchingTags(tagIds, tagIds.iterator());
+    return getRecipeMetadataQuery(query);
+  }
+
+  public Iterable<RecipeMetadata> getRecipesMatchingCreator(String creatorId) {
+    Query query = DBUtils.recipes().whereEqualTo("creatorId", creatorId);
+    return getRecipeMetadataQuery(query);
+  }
+
+  public Iterable<RecipeMetadata> getRecipesSavedBy(String userId) {
+    Iterable<String> saved_Ids = User.savedRecipeIds(userId);
+    return getRecipesMatchingIDs(saved_Ids);
+  }
+
+  public Iterable<RecipeMetadata> getRecipesMatchingIDs(Iterable<String> Ids) {
+    ArrayList<String> idList = new ArrayList<>();
+    Iterator<String> iter = Ids.iterator();
+    while (iter.hasNext()) {
+      idList.add(iter.next());
+    }
+    Query query = DBUtils.recipes().whereIn(Recipe.ID_KEY, idList);
+    return getRecipeMetadataQuery(query);
+  }
 
-    public Comment addComment(Comment newComment) {
-        return null;
+  private Iterable<RecipeMetadata> getRecipeMetadataQuery(Query query) {
+    ApiFuture<QuerySnapshot> querySnapshotFuture = query.get();
+    ArrayList<RecipeMetadata> recipeList = new ArrayList<>();
+    QuerySnapshot querySnapshot = DBUtils.blockOnFuture(querySnapshotFuture);
+
+    if (querySnapshot == null) {
+      return null;
     }
-    public void deleteComment(String Id, String recipeId) {
-        return;
+
+    for (DocumentSnapshot document : querySnapshot.getDocuments()) {
+      RecipeMetadata recipe = document.toObject(RecipeMetadata.class);
+      recipeList.add(recipe);
     }
-    public void editCommentContent(String editedContent) {
-        return;
+    return recipeList;
+  }
+
+  /**
+   * Recursively constructs a query on recipes matching all tags passed in
+   *  eg. DBUtils.recipes().whereEqualTo("tag_N-1", true).whereEqualTo("tag_N-2", true)...
+   */
+  private Query recipesMatchingTags(Iterable<String> tagIds, Iterator<String> iter) {
+    if (iter.hasNext()) {
+      String nextTag = iter.next();
+      return recipesMatchingTags(tagIds, iter).whereEqualTo("tags." + nextTag, true);
+    }
```

Comment:
Shouldn't this be some sort of while loop? As is I think if only works if tagIds has only one tag.

Reply:
Nvm, I was reading this wrong. Good use of recursion.